server/internal/server/server.go

```diff
@@ -101,3 +126,144 @@ func (s *S) Run(ctx context.Context, port int, authConfig config.AuthConfig) err
+func (s *S) RunRoute(ctx context.Context, errCh chan error) error {
+	k8sClient, err := k8s.NewClient()
+	if err != nil {
+		return fmt.Errorf("create K8sClient: %v", err)
+	}
+
+	stopper := make(chan struct{})
+	defer close(stopper)
+	defer runtime.HandleCrash()
+
+	engineLabels := map[string]string{
+		engineLabelKey: engineLabelValue,
+	}
+
+	// initialize the routes before starting the informer.
+	if err := s.refreshRoutes(ctx, namespace, engineLabels); err != nil {
+		return err
+	}
+	s.m.printRoute()
+
+	is, err := k8s.NewInformers(k8sClient, stopper, namespace, engineLabels)
+	if err != nil {
+		return err
+	}
+
+	if err := is.SetPodEventHandlers(ctx, []k8s.EventHandlerWithContext{
+		NewPodEventHandler(s.m, errCh),
+	}); err != nil {
+		return err
+	}
+
+	is.Start()
+	log.Printf("Pod informer started.\n")
+
+	// refresh routes periodically, since a engine may evict models.
+	ticker := time.NewTicker(tickPeriod)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			if err := s.refreshRoutes(ctx, namespace, engineLabels); err != nil {
+				return err
+			}
+			s.m.printRoute()
+		case <-stopper:
+			log.Printf("run routes done.\n")
+			return nil
+		case err := <-errCh:
+			log.Printf("run routes error: %s", err)
+			return err
+		}
+	}
+}
+
+func (s *S) refreshRoutes(ctx context.Context, namespace string, labels map[string]string) error {
+	k8sClient, err := k8s.NewClient()
+	if err != nil {
+		return fmt.Errorf("create K8sClient: %v", err)
+	}
+
+	pods, err := k8sClient.ListPods(ctx, namespace, labels)
+	if err != nil {
+		return err
+	}
+	log.Printf("Refreshing routes, found %d pods\n", len(pods))
+
+	newRoute := newRoute()
+	for _, pod := range pods {
+		ip := pod.Status.PodIP
+		if ip == "" {
+			continue
+		}
+		models, err := s.listModels(ctx, ip)
+		if err != nil {
+			return err
+		}
+		for _, m := range models {
+			newRoute.addRoute(model{id: m}, ip)
+		}
+		newRoute.addServer(ip)
+	}
+	newRoute.printRoute()
+
+	s.m.replace(newRoute)
+	return nil
+}
+
+func (s *S) listModels(ctx context.Context, ip string) ([]string, error) {
+	dest := fmt.Sprintf("%s:%d", ip, s.inferenceManagerEngineInternalGRPCPort)
+	log.Printf("Listing loaded models on %s\n", dest)
+	conn, err := grpc.Dial(dest, grpc.WithTransportCredentials(insecure.NewCredentials()))
+	if err != nil {
+		return nil, err
+	}
+	ec := v1.NewInferenceEngineInternalServiceClient(conn)
+	resp, err := ec.ListModels(ctx, &v1.ListModelsRequest{})
+	if err != nil {
+		return nil, err
+	}
+
+	var modelIDs []string
+	for _, model := range resp.Models {
+		modelIDs = append(modelIDs, model.Id)
+	}
+	return modelIDs, nil
+}
+
+func (s *S) getDestination(ctx context.Context, modelID string) (string, error) {
+	routes := s.m.getRoute(model{id: modelID})
+	if len(routes) != 0 {
+		// TODO(guangrui): handle load balance.
+		return routes[0], nil
+	}
+	return s.pullModel(ctx, modelID)
+}
+
+func (s *S) pullModel(ctx context.Context, modelID string) (string, error) {
```

Comment:
The return value of this function wasn't clear to me. Might make sense to move line 249-253 to `getDestination`.